shared/js/sk_key_utils.js

```diff
@@ -0,0 +1,367 @@
+      form.setAttribute('role', 'dialog');
+      form.tabIndex = -1;
+
+      // We append title if needed
+      if (options.header) {
+        var header = document.createElement('header');
+
+        if (typeof options.header === 'string') {
+          header.textContent = options.header || '';
+        }
+
+        form.appendChild(header);
+      }
+
+      // We append a menu with the list of options
+      var menu = document.createElement('menu');
+      menu.dataset.items = items.length;
+
+      // For each option, we append the item and listener
+      items.forEach(function renderOption(item) {
+        var button = document.createElement('button');
+        button.type = 'button';
+        if (item.label && item.label.length) {
+          button.textContent = item.label || '';
+        } else {
+          return;
+        }
+
+        if (item.icon) {
+          button.style.backgroundImage = 'url("' + item.icon + '")';
+          button.classList.add('icon');
+        }
+
+        // Listener will be removed once we clean up the DOM
+        button.addEventListener('click', function (event) {
+          var clickEvt = new MouseEvent('click', {
+            'view': window,
+            'bubbles': true,
+            'cancelable': true
+          });
+          item.dispatchEvent(clickEvt);
+        });
+
+        menu.appendChild(button);
+      });
+
+      // Appending the action menu to the form
+      form.appendChild(menu);
+
+      // Add cancel key
+      var footer = document.createElement('footer');
+      footer.textContent = sk_options_menu.cancelKeyText;
+      footer.classList.add('sk-cancel-key');
+
+      // TODO: Need to change l10nid, wrong way
+      footer.setAttribute('data-l10n-id',
+              sk_options_menu.cancelKeyl10nId);
+
+      // Handle Close click, #IFDEF check required
+      // for TOUCH UI
+      footer.addEventListener('click', function () {
+        hideOptionsMenu();
+      });
+
+      form.appendChild(footer);
+
+      // Show Menu
+      sk_container.insertBefore(form, sk_container.firstChild);
+      form.classList.add('visible');
+
+      sk_options_menu.open = true;
+    }
+
+    /**
+     * Used for hiding options menu
+     * @example
+     * hideOptionsMenu();
+     *
+     * @param N/A
+     * @return void 0
+     */
+    function hideOptionsMenu() {
+      sk_container
+              .querySelector('form')
+              .remove('visible');
+
+      sk_options_menu.open = false;
+    }
+
+    /**
+     * Used to observe DOM changes
+     * @example
+     * observeDOMChange(element);
+     *
+     * @param {object} target: Target element on which observer will observe
+     * @return void 0
+     */
+    function observeDOMChange(target) {
+      var MutationObserver = window.MutationObserver ||
+              window.WebKitMutationObserver ||
+              window.MozMutationObserver;
+      if (!MutationObserver) {
+        throw Error('MutationObserver not supported');
+      }
```

Comment:
Remember that you're not writing code where cross-browser compatibility is going to matter. This is only going to run in gecko 37 and above.  You can just use MutationObserver without any of the stuff above.